lib/vtls/sectransp.c

```diff
@@ -138,6 +139,474 @@ struct ssl_backend_data {
+             CIPHER_WEAK_NOT_ENCRIPTED),
+  CIPHER_DEF(TLS_ECDH_ECDSA_WITH_RC4_128_SHA,                      /* 0xC002 */
+             CIPHER_WEAK_RC_ENCRIPTION),
+  CIPHER_DEF(TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA,                 /* 0xC003 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA,                  /* 0xC004 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA,                  /* 0xC005 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDHE_ECDSA_WITH_NULL_SHA,                        /* 0xC006 */
+             CIPHER_WEAK_NOT_ENCRIPTED),
+  CIPHER_DEF(TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,                     /* 0xC007 */
+             CIPHER_WEAK_RC_ENCRIPTION),
+  CIPHER_DEF(TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA,                /* 0xC008 */
+             CIPHER_STRONG_ARGUABLY_3DES),
+  CIPHER_DEF(TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,                 /* 0xC009 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,                 /* 0xC00A */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDH_RSA_WITH_NULL_SHA,                           /* 0xC00B */
+             CIPHER_WEAK_NOT_ENCRIPTED),
+  CIPHER_DEF(TLS_ECDH_RSA_WITH_RC4_128_SHA,                        /* 0xC00C */
+             CIPHER_WEAK_RC_ENCRIPTION),
+  CIPHER_DEF(TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA,                   /* 0xC00D */
+             CIPHER_STRONG_ARGUABLY_3DES),
+  CIPHER_DEF(TLS_ECDH_RSA_WITH_AES_128_CBC_SHA,                    /* 0xC00E */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDH_RSA_WITH_AES_256_CBC_SHA,                    /* 0xC00F */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDHE_RSA_WITH_NULL_SHA,                          /* 0xC010 */
+             CIPHER_WEAK_NOT_ENCRIPTED),
+  CIPHER_DEF(TLS_ECDHE_RSA_WITH_RC4_128_SHA,                       /* 0xC011 */
+             CIPHER_WEAK_RC_ENCRIPTION),
+  CIPHER_DEF(TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,                  /* 0xC012 */
+             CIPHER_STRONG_ARGUABLY_3DES),
+  CIPHER_DEF(TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,                   /* 0xC013 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,                   /* 0xC014 */
+             CIPHER_STRONG_ENOUGH),
+  CIPHER_DEF(TLS_ECDH_anon_WITH_NULL_SHA, CIPHER_WEAK_ANON_AUTH),  /* 0xC015 */
+  CIPHER_DEF(TLS_ECDH_anon_WITH_RC4_128_SHA,                       /* 0xC016 */
+             CIPHER_WEAK_ANON_AUTH),
+  CIPHER_DEF(TLS_ECDH_anon_WITH_3DES_EDE_CBC_SHA,                  /* 0xC017 */
+             CIPHER_STRONG_ARGUABLY_3DES),
+  CIPHER_DEF(TLS_ECDH_anon_WITH_AES_128_CBC_SHA,                   /* 0xC018 */
+             CIPHER_WEAK_ANON_AUTH),
+  CIPHER_DEF(TLS_ECDH_anon_WITH_AES_256_CBC_SHA,                   /* 0xC019 */
+             CIPHER_WEAK_ANON_AUTH),
+#endif /* CURL_BUILD_MAC_10_6 || CURL_BUILD_IOS */
```

Comment:
I don't see TLS_DH_anon_WITH_3DES_EDE_CBC_SHA on here. That weak cipher-suite needs to be disabled.

Reply:
It is in lines 293-294, cipher number 0x1B.
Macro CIPHER_DEF_SSLTLS defines cipher with 2 names, one starts with "SSL_" prefix, another with "TLS_", and enum is referenced by SSL_ prefix. Last part we can skip from discussion for now.
Two names are done only for backward compatibility with previous/existing code.
Existing code has two functions `SSLCipherNameForNumber` and `TLSCipherNameForNumber` to return a cipher name by its number. First one is called when negotiated communication protocol is SSL. Second function is called for TLS communication. Result for cipher numbers 0x01 to 0x1B is different, depending on function. First function returns name starting with "SSL_" prefix while second one for the same numbers return name with prefix "TLS_". For example, name of the cipher number 0x01 is "SSL_RSA_WITH_NULL_MD5" when returned by the function `SSLCipherNameForNumber` and the same number mapped to "TLS_RSA_WITH_NULL_MD5" by the second function.
I wanted to keep it the same way and each cipher has 2 names: field `name` points to TLS_ name, field `ssl_name` points to SSL_ name. Those names are used for logging only.
Macro CIPHER_DEF_SSLTLS defines ciphers with both names but its argument is not the real number constant, it is a constant without TLS_ or SSL_ prefix. The macro concatenates prefixes to define both names, as well as concatenates parameter with SSL_ prefix to get the enum number properly. It is a C-style for some code generation, not the best style but I did not find better one.
You did not find `TLS_DH_anon_WITH_3DES_EDE_CBC_SHA` constant but you can find `DH_anon_WITH_3DES_EDE_CBC_SHA` in the new code.

I think better approach would be to simplify all that dance and drop names starting with "SSL_", always return "TLS_" name. I don't know how important backward compatibility for logging. I found that OpenSSL has some trace reporting code and they always print there "TLS_" names. If reviewers are fine with it, I'll simplify the code and make it always returning "TLS_" names, and all constants in the code will be the full constant like `TLS_DH_anon_WITH_3DES_EDE_CBC_SHA` not the `DH_anon_WITH_3DES_EDE_CBC_SHA`. To be more precise, it will be SSL_DH_anon_WITH_3DES_EDE_CBC_SHA, but we can discuss it next time.

Those complications related to macOS SDK constants and support were one of the reasons to create table I added to another file. I had other reasons also. I just want to explain my thoughts and reasons here.
I think it would be better to split this change on 2: one for source code only, another will have only the new table with cipher mappings. The table needs more work to be done.
I very appreciate any ideas and opinions.
Thank you! :)